### DISCUSSION TITLE
Converting clipboard image to base64?

### DISCUSSION THREAD

**Discussion by magpie514** [hluk/CopyQ#2018] [Q&A] [answered]
What would be the best way to convert an image in the clipboard to a base64 string?
I want to do something like 
* Select image in clipboard
* Send to external tool to convert to base64 and add a couple strings around it (markdown formatting for links and data urls)
* Obtain result and add it to main clipboard tab, and select it.

But I'm a bit unsure on how to proceed, particularly the getting the image to an external script. Can anyone help me figure this out?

---
### Calling external script
``` ini
[Command]
Command="
    copyq:
      var img = data(\"image/png\")
      var pth = Dir().homePath() + \"/.local/share/scripts/img2base64.sh\"
      var cmd = execute(pth, null, img)
      if (cmd && cmd.exit_code == 0) {
        var result=str(cmd.stdout)
        tab('clipboard')
        add(result)
        copy(result)
      }"
Icon=\xf03e
InMenu=true
Input=image/png
Name=Base64
```
### Pure copyq
```ini
[Command]
Command="
    copyq:
      var img = data(\"image/png\")
      var result = \"data:image/png;base64,\" + toBase64(img)
      tab('clipboard')
      add(result)
      copy(result)"
Enable=false
Icon=\xf03e
InMenu=true
Input=image/png
Name=Base64
```

Can probably be optimized, but does the job.

**Comment by dbareis** [comment]
There is already a function to convert to base 64 "[toBase64()](https://copyq.readthedocs.io/en/latest/scripting-api.html#dialog:~:text=setSelectedItemsData()-,toBase64,-(data))"

To paste result:

```
B64 = toBase64(...)
copy(B64)
paste()
```

**Reply by magpie514** [reply]
Ah, great, that simplifies things a lot, thank you.
Do you know how I'd grab the image data to encode?

**Comment by dbareis** [comment, marked as answer]
You said it is in the clipboard, read the clipboard to get the data (click F4 on item to see the available formats and their values).

**Reply by magpie514** [reply]
Yeah, that's what I'm asking about how to do, how to get the currently selected image and get its `image/png` content.

**Reply by magpie514** [reply]
Ah! Figured it out. 
Turns out it's just `data("image/png")`, examples seemed to need to iterate among possible multiple items but seems the current object is somewhat present in scripts without having to assign it to a variable or anything first, that's what was throwing me off.
